ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/streaming/QueryStreamWriter.java

```diff
@@ -61,82 +65,62 @@
-              out.write("\n".getBytes());
-              out.flush();
-            }
-          }
-        }
-      } catch (EOFException exception) {
-        // The user has terminated the connection; we can stop writing
-      } catch (InterruptedException exception) {
-        // The most likely cause of this is the server shutting down. Should just try to close
-        // gracefully, without writing any more to the connection stream.
-        log.warn("Interrupted while writing to connection stream");
-      } catch (Throwable exception) {
-        log.error("Exception occurred while writing to connection stream: ", exception);
-        synchronized (out) {
-          out.write("\n".getBytes());
-          if (exception.getCause() instanceof KsqlException) {
-            new ObjectMapper().writeValue(out, new StreamedRow(exception.getCause()));
-          } else {
-            new ObjectMapper().writeValue(out, new StreamedRow(exception));
-          }
           out.write("\n".getBytes());
           out.flush();
         }
-      }
-
-      if (rowWriterThread.isAlive()) {
-        try {
-          rowWriterThread.interrupt();
-          rowWriterThread.join();
-        } catch (InterruptedException exception) {
-          log.warn(
-              "Failed to join row writer thread; setting to daemon to avoid hanging on shutdown"
-          );
-          rowWriterThread.setDaemon(true);
+        if (streamsException != null) {
+          throw streamsException;
         }
       }
+    } catch (EOFException exception) {
+      // The user has terminated the connection; we can stop writing
+    } catch (InterruptedException exception) {
+      // The most likely cause of this is the server shutting down. Should just try to close
+      // gracefully, without writing any more to the connection stream.
+      log.warn("Interrupted while writing to connection stream");
+    } catch (Throwable exception) {
+      log.error("Exception occurred while writing to connection stream: ", exception);
+      out.write("\n".getBytes());
+      if (exception.getCause() instanceof KsqlException) {
+        new ObjectMapper().writeValue(out, new StreamedRow(exception.getCause()));
+      } else {
+        new ObjectMapper().writeValue(out, new StreamedRow(exception));
+      }
+      out.write("\n".getBytes());
+      out.flush();
 
     } finally {
       queryMetadata.getKafkaStreams().close(100L, TimeUnit.MILLISECONDS);
       queryMetadata.getKafkaStreams().cleanUp();
     }
   }
 
+  void write(OutputStream output, GenericRow row) throws IOException {
+    objectMapper.writeValue(output, new StreamedRow(row));
+    output.write("\n".getBytes());
+    output.flush();
+  }
+
   private class StreamsExceptionHandler implements Thread.UncaughtExceptionHandler {
+
     @Override
     public void uncaughtException(Thread thread, Throwable exception) {
-      streamsException.compareAndSet(null, exception);
+      streamsException = exception;
```

Comment:
if we get a StreamsException the only sensible thing to do really is shutdown

Reply:
it gets passed back to the stream loop above,  that loop will rethrow and cleanup